TulparAudio/internal/source/SourceCollection.cpp

```diff
@@ -0,0 +1,1097 @@
+    alGetSourcei(static_cast<ALuint>(source), AL_BUFFERS_PROCESSED, &alQueueIndex);
+
+    alErr = alGetError();
+
+    if (AL_NO_ERROR != alErr)
+    {
+        LOG_AUDIO->Warning("Source #{}: get buffer queue index: {:#x}", source, alErr);
+        alQueueIndex = 0;
+    }
+
+    return alQueueIndex;
+}
+
+bool SourceCollection::QueueSourceBuffers(SourceHandle source, std::vector<audio::Buffer> const& buffers)
+{
+    LOG_AUDIO->Debug("Source #{}: set buffer queue[{}]", source, buffers.size());
+
+    std::vector<ALuint> tmp(buffers.size(), 0);
+
+    std::transform(buffers.begin(), buffers.end(), tmp.begin(),
+        [](audio::Buffer const& buffer) -> ALuint
+        {
+            return static_cast<ALuint>(*(buffer.GetSharedHandle()));
+        }
+    );
+
+    // clear error state
+    ALenum alErr = alGetError();
+
+    alSourceQueueBuffers(static_cast<ALuint>(source), tmp.size(), tmp.data());
+
+    alErr = alGetError();
+
+    if (AL_NO_ERROR == alErr)
+    {
+        using namespace std::chrono_literals;
+
+        std::vector<BufferHandle>& queue = m_sourceQueuedBuffers[source];
+        queue.reserve(buffers.size());
+
+        Meta& meta = m_sourceMeta[source];
+
+        meta.activeSampleCount = 0;
+        meta.activeTotalDuration = 0ns;
+
+        for (audio::Buffer const& buffer : buffers)
+        {
+            meta.activeSampleCount += buffer.GetSampleCount();
+            meta.activeTotalDuration += buffer.GetDuration();
+
+            queue.push_back(*(buffer.GetSharedHandle()));
+        }
+
+        m_sourceBuffers[source] = *(audio::Buffer().GetSharedHandle());
+    }
+    else
+    {
+        LOG_AUDIO->Warning("Source #{}: set buffer queue[{}]: {:#x}", source, tmp.size(), alErr);
+    }
+
+    return AL_NO_ERROR == alErr;
+}
+
+void SourceCollection::ResetSource(SourceHandle source)
+{
+    LOG_AUDIO->Debug("Source #{}: reset", source);
+
+    Reclaim(source);
+    SetSourceStaticBuffer(source, *(audio::Buffer().GetSharedHandle()));
+
+    m_sourceMeta.erase(source);
+    m_sourceBuffers.erase(source);
+    m_sourceQueuedBuffers.erase(source);
+}
+
+bool SourceCollection::PlaySource(SourceHandle source)
+{
+    LOG_AUDIO->Debug("Source #{}: play", source);
+
+    // clear error state
+    ALenum alErr = alGetError();
+
+    alSourcePlay(static_cast<ALuint>(source));
+
+    alErr = alGetError();
+
+    if (AL_NO_ERROR != alErr)
+    {
+        LOG_AUDIO->Warning("Source #{}: play: {:#x}", source, alErr);
+    }
+
+    return AL_NO_ERROR == alErr;
+}
+
+bool SourceCollection::StopSource(SourceHandle source)
+{
+    LOG_AUDIO->Debug("Source #{}: stop", source);
+
+    // clear error state
+    ALenum alErr = alGetError();
+
+    alSourceStop(static_cast<ALuint>(source));
+
+    alErr = alGetError();
+
+    if (AL_NO_ERROR != alErr)
+    {
+        LOG_AUDIO->Warning("Source #{}: stop: {:#x}", source, alErr);
+    }
+
+    return AL_NO_ERROR == alErr;
+}
+
+bool SourceCollection::RewindSource(SourceHandle source)
```

Comment:
Shouldn't it be const and static?

Reply:
Answered in https://github.com/Godlike/Tulpar/pull/6#discussion_r244268061